File: 766-flatten-a-multilevel-doubly-linked-list/flatten-a-multilevel-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* solve(Node* head)
    {
        Node* curr=head;
        Node* tail=head;
        while(curr)
        {
            if(curr->child)
            {
                Node* childNode=curr->child;
                curr->child=NULL;

                Node* childTail=solve(childNode);

                childTail->next=curr->next;
                if(curr->next)
                curr->next->prev=childTail;

                curr->next=childNode;
                childNode->prev=curr;

                tail=childTail;
            }

            tail=curr;
            curr=curr->next;
        }
        return tail;
    }

    Node* flatten(Node* head) 
    {
        if(!head)
        return head;

        solve(head);
        return head;
    }

};

File: 766-flatten-a-multilevel-doubly-linked-list/flatten-a-multilevel-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/

class Solution {
    public Node flatten(Node head) {
        Node curr = head;
        while(curr != null){
            if(curr.child != null){

                Node nextNode = curr.next;
                Node tail = curr.child;

                curr.next = curr.child;
                curr.child.prev = curr;
                curr.child = null;

                while(tail.next != null){
                    tail = tail.next;
                }

                tail.next = nextNode;
                if(nextNode != null){
                    nextNode.prev = tail;
                }
            }
            curr = curr.next;
        }
        return head;
    }
}